hvantk/ptm/constraint_expression.py

```
@@ -0,0 +1,359 @@
+    return df
+
+
+def _load_from_tabular(path: str) -> pd.DataFrame:
+    """Load a pre-aggregated gene x group matrix from disk."""
+    ext = os.path.splitext(path)[1].lower()
+    logger.info("Loading tabular expression matrix from %s (ext=%s)", path, ext)
+
+    if ext in {".parquet", ".pq"}:
+        df = pd.read_parquet(path)
+    elif ext in {".pkl", ".pickle"}:
+        df = pd.read_pickle(path)
```

Ruff (0.15.9)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)